File: go/sort/quick_sort.go
package main

import "fmt"
import "time"
import "math/rand"


func quick_sort(array []int) {
	if len(array) <= 1 {
		return
	}
	sep := partition(array)
	quick_sort(array[:sep])
	quick_sort(array[sep + 1:])
}


func partition(array []int) int {
	length := len(array)
	middle := rand.Intn(length)
	array[middle], array[length - 1] = array[length - 1], array[middle]
	i := -1
	for j := 0; j < length - 1; j ++ {
		if array[j] < array[length - 1] {
			i ++
			array[i], array[j] = array[j], array[i]
		}
	}
	i ++
	array[i], array[length - 1] = array[length - 1], array[i]

	return i
}


func main() {
	rand.Seed(time.Now().UTC().UnixNano())
	test_array := make([]int, 20)
	for i := 0; i < 20; i ++ {
		test_array[i] = rand.Intn(20)
	}
	fmt.Printf("Before sort: ")
	fmt.Println(test_array)
	quick_sort(test_array)
	fmt.Printf("After sort: ")
	fmt.Println(test_array)
}

File: python/sort/quick_sort.py
import random


def quick_sort(array):
    if len(array) <= 1:
        return array
    middle = partition(array)
    left = quick_sort(array[0:middle])
    right = quick_sort(array[middle + 1:len(array)])
    array[:] = left + [array[middle]] + right
    return array


def partition(array):
    i = -1
    choose_middle(array)
    for j in xrange(len(array) - 1):
        if array[j] < array[len(array) - 1]:
            i += 1
            array[i], array[j] = array[j], array[i]
    i += 1
    array[len(array) - 1], array[i] = array[i], array[len(array) - 1]
    return i


def choose_middle(array):
    middle = random.randint(0, len(array) - 1)
    array[middle], array[len(array) - 1] = array[len(array) - 1], array[middle]
